src/storage/src/hummock/compactor/compactor_runner.rs

```diff
@@ -356,151 +317,52 @@ pub async fn compact(
             .filter(|table_id| existing_table_ids.contains(table_id)),
     );
-    let multi_filter_key_extractor = match filter_key_extractor_manager
-        .acquire(compact_table_ids.clone())
-        .await
+
+    let multi_filter_key_extractor = match build_filter_key_extractor(
+        &compact_task,
+        filter_key_extractor_manager.clone(),
+        &compact_table_ids,
+    )
+    .await
     {
-        Err(e) => {
-            tracing::error!(error = %e.as_report(), "Failed to fetch filter key extractor tables [{:?}], it may caused by some RPC error", compact_task.existing_table_ids);
+        Some(multi_filter_key_extractor) => multi_filter_key_extractor,
+        None => {
             let task_status = TaskStatus::ExecuteFailed;
             return (
                 compact_done(compact_task, context.clone(), vec![], task_status),
                 None,
             );
         }
-        Ok(extractor) => extractor,
     };
 
-    if let FilterKeyExtractorImpl::Multi(multi) = &multi_filter_key_extractor {
-        let found_tables = multi.get_existing_table_ids();
-        let removed_tables = compact_table_ids
-            .iter()
-            .filter(|table_id| !found_tables.contains(table_id))
-            .collect_vec();
-        if !removed_tables.is_empty() {
-            tracing::error!("Failed to fetch filter key extractor tables [{:?}. [{:?}] may be removed by meta-service. ", compact_table_ids, removed_tables);
-            let task_status = TaskStatus::ExecuteFailed;
-            return (
-                compact_done(compact_task, context.clone(), vec![], task_status),
-                None,
-            );
-        }
-    }
-
-    let multi_filter_key_extractor = Arc::new(multi_filter_key_extractor);
-    let has_tombstone = compact_task
-        .input_ssts
-        .iter()
-        .flat_map(|level| level.table_infos.iter())
-        .any(|sst| sst.range_tombstone_count > 0);
-    let has_ttl = compact_task
-        .table_options
-        .iter()
-        .any(|(_, table_option)| table_option.retention_seconds.is_some_and(|ttl| ttl > 0));
     let mut task_status = TaskStatus::Success;
-    // skip sst related to non-existent able_id to reduce io
-    let sstable_infos = compact_task
-        .input_ssts
-        .iter()
-        .flat_map(|level| level.table_infos.iter())
-        .filter(|table_info| {
-            let table_ids = &table_info.table_ids;
-            table_ids
-                .iter()
-                .any(|table_id| existing_table_ids.contains(table_id))
-        })
-        .cloned()
-        .collect_vec();
-    let compaction_size = sstable_infos
-        .iter()
-        .map(|table_info| table_info.file_size)
-        .sum::<u64>();
-    let all_ssts_are_blocked_filter = sstable_infos
-        .iter()
-        .all(|table_info| table_info.bloom_filter_kind() == BloomFilterType::Blocked);
+    let optimize_by_copy_block = optimize_by_copy_block(&compact_task, &context);
```

Comment:
Given that we already pre-calculated `optimize_by_copy_block` prior to entering the `compact` function in `calculate_task_parallelism`. I think we don't need to re-compute it again here. How about storing the `parallelism` and `optimize_by_copy_block` in `CompactorContext` so that they can be directly used here.

Reply:
Can we tolerate this issue in this PR? From my perspective, `Context` does not belong to a task, but to a virtual compactor runner. However, I don't have a better solution for the time being.

Reply:
I see. It is okay. This is not a big deal.